### PR TITLE
ARROW-8732: [C++] Add basic cancellation API

### PR DESCRIPTION
**Pull request by pitrou**
In this model, a StopSource is instantiated by the consumer, which passes a corresponding StopToken to producer API(s).

### CODE REVIEW

File: cpp/src/arrow/util/future.h [hunk 1]
```diff
@@ -515,15 +548,35 @@ class ARROW_MUST_USE_TYPE Future {
...
+        void operator()() { fut->SetResult(std::move(res)); }
+      };
+      impl_->MarkFinished(state, ResultSetter{this, std::move(res)});
     } else {
-      impl_->MarkFailed();
+      // The future cannot be cancelled concurrently as CancelOn()
+      // was not called.
+      SetResult(std::move(res));
+      impl_->MarkFinished(state);
     }
```
Comment (pitrou) [inline cpp/src/arrow/util/future.h:565]: Unfortunately, having the ability to cancel from one thread while marking a future finished from another thread introduces some execution costs inside the Future class. Here, we see that if some cancellation may occur, we must synchronize result writing inside `FutureImpl::MarkFinished`.

cc @westonpace

Comment (westonpace) [reply cpp/src/arrow/util/future.h:565]: We already had a lock so if I understand correctly you are pointing out that the critical section is getting larger (to include the result writing)?

For the waiters it is unlikely to be a big issue.  They are waiting already so they shouldn't be too bothered by a bit more waiting.

For AddCallback/TryAddCallback contention it would have to be a situation where the future the callback being added to was truly asynchronous (spawned work in another thread).  If it was a purely synchronous future then mark finished would be finished being called before AddCallback is called.  So in this case the person calling AddCallback has already accepted they might be adding a callback early and accepting the cost of their task being added to the scheduler and losing their CPU caches.  A little blocking (especially since it results in them being able to run their callback immediately) should already be in the range of costs they are willing to accept.

In other words, you already should not be calling AddCallback on a potentially unfinished future in a performance critical section of code.

Comment (pitrou) [reply cpp/src/arrow/util/future.h:565]: No, I'm pointing out that I'm giving a callable to `FutureImpl::MarkFinished` instead of calling it directly.

**Comment by pitrou** [comment]
Hmm, perhaps instead of complicating the Future implementation, it would be better to make the Executor responsible for setting the stop callback.

File: cpp/src/arrow/util/future.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/util/future.h]: Ok, I removed this and other changes to `future.h`.

**Comment by pitrou** [comment]
@ursabot crossbow submit -g python

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

**Comment by pitrou** [comment]
@westonpace @bkietz Feel free to review.

File: cpp/src/arrow/util/cancel_test.cc [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+#ifndef _WIN32
+static util::optional<StopSource> signal_stop_source;
+
+static void signal_handler(int signum) {
+  signal_stop_source->RequestStopFromSignal(signum);
+}
+
+static void SetITimer(double seconds) {
```
Comment (bkietz) [inline cpp/src/arrow/util/cancel_test.cc:128]:
```suggestion
// SIGALRM will be received once after the specified wait
static void SetITimer(double seconds) {
```

File: cpp/src/arrow/csv/reader.cc [hunk 3]
```diff
@@ -833,9 +843,10 @@ class AsyncThreadedTableReader
   AsyncThreadedTableReader(MemoryPool* pool, std::shared_ptr<io::InputStream> input,
                            const ReadOptions& read_options,
                            const ParseOptions& parse_options,
-                           const ConvertOptions& convert_options, Executor* cpu_executor,
-                           Executor* io_executor)
-      : BaseTableReader(pool, input, read_options, parse_options, convert_options),
+                           const ConvertOptions& convert_options, StopToken stop_token,
+                           Executor* cpu_executor, Executor* io_executor)
```
Comment (bkietz) [inline cpp/src/arrow/csv/reader.cc:847]: Seems we might prefer to rewrite this constructor to take an IOContext

**Review by bkietz — CHANGES_REQUESTED** [review]
This is looking great!

A few comments:

File: cpp/src/arrow/util/thread_pool_test.cc [hunk 4]
```diff
@@ -137,30 +136,47 @@ class TestThreadPool : public ::testing::Test {
     return *ThreadPool::Make(threads);
   }
 
-  void SpawnAdds(ThreadPool* pool, int nadds, AddTaskFunc add_func) {
-    AddTester add_tester(nadds);
+  void SpawnAdds(ThreadPool* pool, int nadds, AddTaskFunc add_func,
+                 StopToken stop_token = StopToken::Unstoppable(),
+                 StopSource* stop_source = nullptr) {
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/thread_pool_test.cc:141]: Passing the token with not the source tests a source that's not stopped, so it's deliberate. I could try to add overloads though.

File: cpp/src/arrow/util/cancel.cc [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  if (!impl_->requested_.load()) {
+    return Status::OK();
+  }
+
+  std::lock_guard<std::mutex> lock(impl_->mutex_);
+  if (impl_->cancel_error_.ok()) {
+    auto signum = impl_->requested_.load();
+    DCHECK_GT(signum, 0);
+    impl_->cancel_error_ = internal::CancelledFromSignal(signum, "Operation cancelled");
+  }
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/cancel.cc:91]: If you call Poll() twice, then the cancel_error corresponding to the signal will have been cached.

File: cpp/src/arrow/util/cancel.cc [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+Result<StopSource*> SetSignalStopSource() {
+  if (g_signal_stop_state) {
+    return Status::Invalid("Signal stop source already set up");
+  }
+  internal::atomic_store(&g_signal_stop_state, std::make_shared<SignalStopState>());
+  return &g_signal_stop_state->stop_source;
+}
+
+void ResetSignalStopSource() {
+  internal::atomic_store(&g_signal_stop_state, std::shared_ptr<SignalStopState>{});
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/cancel.cc:151]: Good point, will change.

File: cpp/src/arrow/csv/reader.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/csv/reader.cc:847]: We must support the legacy `TableReader::Make` taking both a `MemoryPool*` and an `IOContext`.

File: cpp/src/arrow/util/task_group.cc [hunk 7]
```diff
@@ -67,34 +75,54 @@ class SerialTaskGroup : public TaskGroup {
...
-          },
-          std::move(self), std::move(task)));
+            self_->UpdateStatus(std::move(st));
+          }
+          self_->OneTaskDone();
+        }
+
+        std::shared_ptr<ThreadedTaskGroup> self_;
+        FnOnce<Status()> task_;
+        StopToken stop_token_;
```
Comment (pitrou) [inline cpp/src/arrow/util/task_group.cc:121]: Note to self: can use `self_->stop_token_` instead.

File: cpp/src/arrow/util/thread_pool.cc [hunk 8]
```diff
@@ -91,12 +101,20 @@ static void WorkerLoop(std::shared_ptr<ThreadPool::State> state,
...
-        std::move(task)();
+        if (!stop_token->IsStopRequested()) {
+          std::move(task.callable)();
+        } else {
+          if (task.stop_callback) {
+            std::move(task.stop_callback)(stop_token->Poll());
+          }
+        }
+        ARROW_UNUSED(std::move(task));  // release resources before waiting for lock
+        lock.lock();
```
Comment (pitrou) [inline cpp/src/arrow/util/thread_pool.cc:116]: Note to self: better to re-lock outside of scope? (to let `task` be destroyed)

File: python/pyarrow/_csv.pyx [hunk 9]
```diff
@@ -34,8 +34,8 @@ from pyarrow.lib cimport (check_status, Field, MemoryPool, Schema,
                           pyarrow_unwrap_batch, pyarrow_unwrap_table,
                           pyarrow_wrap_schema, pyarrow_wrap_table,
                           pyarrow_wrap_data_type, pyarrow_unwrap_data_type,
-                          Table, RecordBatch)
-from pyarrow.lib import frombytes, tobytes
+                          Table, RecordBatch, StopToken)
```
Comment (westonpace) [inline python/pyarrow/_csv.pyx:37]: Is there a JIRA to expose `StopSource` to python at some point?  There are reasons other than signals someone may want to cancel an operation.  For example, a GUI-based application may have a cancel button.  A web server may want to cancel if the TCP connection for some analysis request is lost.

**Review by westonpace — COMMENTED** [review]
I learned a lot about signals just reading this :smile:

File: cpp/src/arrow/csv/reader.cc [hunk 10]
```diff
@@ -848,6 +859,7 @@ class AsyncThreadedTableReader
   }
 
   Status Init() override {
+    // TODO pass stop token to input stream iterator?
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/csv/reader.cc:862]: Indeed, the AWS SDK doesn't support cancellation. 

Agreed about the readahead part. This can be done later. Thanks!

File: cpp/src/arrow/csv/reader.cc [hunk 11]
```diff
@@ -934,22 +946,34 @@ class AsyncThreadedTableReader
 Result<std::shared_ptr<TableReader>> MakeTableReader(
     MemoryPool* pool, io::IOContext io_context, std::shared_ptr<io::InputStream> input,
     const ReadOptions& read_options, const ParseOptions& parse_options,
-    const ConvertOptions& convert_options) {
+    const ConvertOptions& convert_options, StopToken stop_token) {
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/csv/reader.cc:949]: Agreed, I didn't think too much about this. Will change.

File: cpp/src/arrow/util/cancel.cc [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+  DCHECK(!st.ok());
+  if (!impl_->requested_) {
+    impl_->requested_ = -1;
+    impl_->cancel_error_ = std::move(st);
+  }
+}
+
+void StopSource::RequestStopFromSignal(int signum) {
+  // Only async-signal-safe code allowed here
+  impl_->requested_.store(signum);
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/cancel.cc:66]: No, we need the mutex to protect `cancel_error_` (which we cannot store atomically).

File: cpp/src/arrow/util/cancel.cc [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+  StopSource stop_source;
+  std::vector<SavedSignalHandler> saved_handlers;
+};
+
+std::shared_ptr<SignalStopState> g_signal_stop_state;
+
+void HandleSignal(int signum) {
+  ReinstateSignalHandler(signum, &HandleSignal);
+  std::shared_ptr<SignalStopState> state = internal::atomic_load(&g_signal_stop_state);
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/cancel.cc:134]: Ouch, I think you're right.

File: cpp/src/arrow/util/cancel.h [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+/// EXPERIMENTAL
+class ARROW_EXPORT StopSource {
+ public:
+  StopSource();
+  ~StopSource();
+
+  // Consumer API (the side that stops)
+  void RequestStop();
+  void RequestStop(Status error);
+  void RequestStopFromSignal(int signum);
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/cancel.h:45]: It can indeed be useful if the application catches signals by itself.

File: python/pyarrow/_csv.pyx (hunk 9 above)
Comment (pitrou) [reply python/pyarrow/_csv.pyx:37]: No JIRA yet. We can open one, or do it when needed.

**Comment by pitrou** [comment]
I pushed an improvement for the deallocation in signal handler problem, but it's still not 100% safe. I think solving it entirely would need a lock-free doubly-linked list.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

**Comment by westonpace** [comment]
Abseil (seems to be Google's equivalent of Folly) has an async-signal-safe spin lock.  That might be easier than a doubly linked list.  However, it doesn't seem to be easily vendored (just based on the fact that it has 10 abseil includes) :(

https://github.com/abseil/abseil-cpp/blob/master/absl/base/internal/spinlock.h

I'd agree that it seems too unlikely to worry about.

**Comment by pitrou** [comment]
How would a spinlock help the deallocation issue?

**Comment by westonpace** [comment]
The handler would obtain the lock before calling `atomic_load` on the pointer and reset the pointer before it releases the lock.  Any call to `ResetSignalStopSource` would grab the lock before modifying the pointer.  This should ensure that the pointer obtained by the load cannot go out of scope while the signal handler references it and thus would never have to free.

**Comment by pitrou** [comment]
Hmm, but resetting the pointer may still deallocate the underlying object... which is certainly not async-signal-safe.

**Comment by westonpace** [comment]
It could only deallocate it if the global reference was changed and all attempts to change the global reference should go through the same spinlock.  The signal handler should always be given the "second reference" no matter what.

Or, since you're using spinlocks to guard all changes to the global reference, you could just change the global reference to a plain old pointer which might make it more obvious.

**Comment by pitrou** [comment]
Ah, I see: no need to take a strong reference inside the handler since the spinlock guarantees the global reference isn't mutated. That said, the signal-safe spinlock apparently requires one to block signals.

**Comment by westonpace** [comment]
Yep.  I think what you have, with the comment, is fine.
